### PR TITLE
New arm arm cr

### PR DESCRIPTION
**Pull request by YHerm**
the misc and jarRepo are in git ignore. I dont know why its here🤗

### CODE REVIEW

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+import edu.greenblitz.utils.GBCommand;
+import edu.wpi.first.wpilibj2.command.Command;
+import edu.wpi.first.wpilibj2.command.SequentialCommandGroup;
+
+import java.util.LinkedList;
+import java.util.function.Supplier;
+
+import static edu.greenblitz.tobyDetermined.Nodesssss.NodeBase.CreateCurrents.system1CurrentNode;
+
+public class GeneratePath implements Supplier<Command> {
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java:14]: Try finding a more descriptive name for the class

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        targetNode = endNode;
+    }
+
+    public Command get() {
+        SystemsState start = system1CurrentNode.getCurrentNode();
+        LinkedList<SystemsState> path = AStar.getPath(start, targetNode);
+        GBCommand[] nodeCommands = new GBCommand[path.size() - 1];
+        for (int i = 0; i < path.size() - 1; i++) {
+            nodeCommands[i] = new StraightLineMotion(path.get(i), path.get(i + 1));
+        }
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java:28]: Not that bad but try finding a cleaner way to do this

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+            combinedVelocity = TO_BELLY_VELOCITY;
+
+        else {
+            double length1 = startingNode.getExtendPosition() + STARTING_LENGTH;
+            double length2 = endNode.getExtendPosition() + STARTING_LENGTH;
+            double distance = GBMath.distance(
+                    GBMath.convertToCartesian(length1, startingNode.getAnglePosition()),
+                    GBMath.convertToCartesian(length2, endNode.getAnglePosition())
+            );
+            combinedVelocity = STATIC_COMBINED_VELOCITY * GBMath.sigmoid(distance, 1, 5.9, 0.3);//numbers that work well
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java:60]: Magic numbers

Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java:60]: Is that comment necessary?

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java [hunk 4]
```diff
@@ -0,0 +1,146 @@
...
+    public void initialize() {
+        if (extender.getLength() + endNode.getExtendPosition() <= DISTANCE_NOT_STRAIGHT_LINE)
+            combinedVelocity = ONLY_ELBOW_VELOCITY;
+
+        else if (endState.equals(SystemsState.ARM_GROUND))
+            combinedVelocity = TO_BELLY_VELOCITY;
+
+        else {
+            double length1 = startingNode.getExtendPosition() + STARTING_LENGTH;
+            double length2 = endNode.getExtendPosition() + STARTING_LENGTH;
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java:55]: Try finding better names for these `double`s

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java [hunk 5]
```diff
@@ -0,0 +1,146 @@
...
+    public double calculateAngularVelocity(double startVelocity) {
+        double signOfAngle = Math.signum(endNode.getAnglePosition() - elbow.getAngleRadians());
+        double magnitudeOfVelocity = startVelocity / (extender.getLength() + STARTING_LENGTH);
+
+        magnitudeOfVelocity = limit(magnitudeOfVelocity, MAX_ANGULAR_VELOCITY);
+
+        return signOfAngle * Math.abs(magnitudeOfVelocity);
+    }
+
+    public double limit(double v, double maxMagnitude) {
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java:92]: What is v?

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java [hunk 6]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    public double limit(double v, double min, double max) {
+        return Math.min(max, Math.max(min, v));
+    }
+
+    public void moveArm() {
+        double startLength = extender.getLength() + STARTING_LENGTH;
+        double endLength = endNode.getExtendPosition() + STARTING_LENGTH;
+        double gamma = endNode.getAnglePosition() - elbow.getAngleRadians();
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java:103]: [gamma](https://www.mitgaisim.idf.il/%D7%AA%D7%A4%D7%A7%D7%99%D7%93%D7%99%D7%9D/%D7%9E%D7%A1%D7%9C%D7%95%D7%9C-%D7%92%D7%90%D7%9E%D7%90-%D7%A1%D7%99%D7%99%D7%91%D7%A8/)?

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java [hunk 7]
```diff
@@ -0,0 +1,146 @@
...
+            double length1 = startingNode.getExtendPosition() + STARTING_LENGTH;
+            double length2 = endNode.getExtendPosition() + STARTING_LENGTH;
+            double distance = GBMath.distance(
+                    GBMath.convertToCartesian(length1, startingNode.getAnglePosition()),
+                    GBMath.convertToCartesian(length2, endNode.getAnglePosition())
+            );
+            combinedVelocity = STATIC_COMBINED_VELOCITY * GBMath.sigmoid(distance, 1, 5.9, 0.3);//numbers that work well
+        }
+
+        Logger.getInstance().recordOutput("Arm/TargetPose3D/endNode", ArmSimulation.getArmPosition(endNode.getExtendPosition(), (endNode.getAnglePosition())));
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java:63]: Magic string

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java [hunk 8]
```diff
@@ -0,0 +1,146 @@
...
+    }
+
+    public static double getRatioBetweenAngleAndLength(double sideA, double sideB, double gamma) {
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java:66]: Could this be moved outside of the class? Doesn't seem relevant to the command itself and the file is kinda long

In general, I see a few functions that are a bit misplaced

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java [hunk 9]
```diff
@@ -0,0 +1,146 @@
...
+
+    @Override
+    public void end(boolean interrupted) {
+        if (interrupted) {
+            system1MidNode.setMidNode(startState, endState);
+            system1CurrentNode.setCurrentNode(SystemsState.MID_NODE_1);
+        } else
+            system1CurrentNode.setCurrentNode(endState);
+    }
+}
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java:146]: File ~could~ should be shorter

Comment (Stormageddon37) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java:146]: "The first rule of classes is that they should be small. The second rule of classes is that they should be smaller than that.".

File: src/main/java/edu/greenblitz/utils/GBMath.java [hunk 10]
```diff
@@ -1,20 +1,92 @@
...
+    /**
+     * Takes an x and y position and speeds, and radius of polar conversion and converts it into polar speeds
+     *
+     * @param xPos      the x coordinate of the point
+     * @param yPos      the y coordinate of the point
+     * @param xVelocity the x Velocity of the point
+     * @param yVelocity the y Velocity of the point
+     * @return a Pair which houses the result in this format (length,angle)
+     */
+    public static Pair<Double, Double> convertToPolarSpeeds(double xPos, double yPos, double xVelocity, double yVelocity) {
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/utils/GBMath.java:52]: [Pos](https://www.urbandictionary.com/define.php?term=POS)?

File: src/main/java/edu/greenblitz/utils/GBMath.java [hunk 11]
```diff
@@ -1,20 +1,92 @@
...
+
+        double lengthVelocity = xPos * (xVelocity) + yPos * (yVelocity);
+        lengthVelocity /= distance;
+
+        double angularVelocity = xPos * yVelocity - yPos * xVelocity;
+        angularVelocity /= distance;
+        return new Pair<>(lengthVelocity, angularVelocity);
+    }
+
+    public static double distance(Translation2d p1, Translation2d p2) {
```
Comment (Stormageddon37) [inline src/main/java/edu/greenblitz/utils/GBMath.java:64]: p1?

The McLaren P1 produced by McLaren Automotive, a supercar with a mid-engine layout with rear-wheel drive and a roof structure safety cage concept called MonoCage?

image

How is that related to "new arm arm cr"? (GREAT name by the way)

Comment (Stormageddon37) [reply src/main/java/edu/greenblitz/utils/GBMath.java:64]: Also `distance` is a bit of a vague name

**Review by Stormageddon37 — CHANGES_REQUESTED** [review]

File: src/main/java/edu/greenblitz/tobyDetermined/RobotMap.java [hunk 12]
```diff
@@ -379,7 +383,8 @@ public static class Elbow {
             public static final int RESET_MEDIAN_SIZE = 7;
             public static final double GEAR_RATIO = 1;
 
-            public static final double kS = 0.133000000000002;
+            public static final double kS = 0.1;
+            public static final double kV = 1.9;
```
[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/edu/greenblitz/tobyDetermined/RobotMap.java:387]: yes it is

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java (hunk 5 above)
Comment (YHerm) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java:92]: value can be anything you want to limit, I changed name to value

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java (hunk 6 above)
Comment (YHerm) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java:103]: angleBetween

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java (hunk 8 above)
Comment (YHerm) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/StraightLineMotion.java:66]: GBMath?

File: src/main/java/edu/greenblitz/tobyDetermined/RobotMap.java [hunk 13]
```diff
@@ -307,6 +307,8 @@ public static class Extender {
             public static final SparkMaxLimitSwitch.Type SWITCH_TYPE = SparkMaxLimitSwitch.Type.kNormallyClosed;
             public static final double MAX_LENGTH_IN_ROBOT = 0.4;
             public static double MAX_ENTRANCE_LENGTH = 0.054;
+            public static final double DISTANCE_NOT_STRAIGHT_LINE = 0.1;
```
[the remark this answers is not in the document]
Comment (YHerm) [reply src/main/java/edu/greenblitz/tobyDetermined/RobotMap.java:310]: STRAIGHT_LINE_UNAVAILABLE_LENGTH?

File: src/main/java/edu/greenblitz/utils/GBMath.java (hunk 11 above)
Comment (YHerm) [reply src/main/java/edu/greenblitz/utils/GBMath.java:64]: Changed to point.
About distance the function is doing the distance equation so i think its a fair name.

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java (hunk 2 above)
Comment (YHerm) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java:28]: I couldn’t think about a way. Do you have a suggestion?

Comment (Stormageddon37) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java:28]: Google "IntStream"

Comment (Stormageddon37) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java:28]: Even using a foreach loop would be better

Comment (YHerm) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java:28]: instead of use IntStream i can do 
path.foreach( state -> ... )

butt i cant foreach here. i will need a prevoius variable and instead of array use list or create counter and thats wayyyyy less cleaner. unless you have an idea how can i know in which index of the array i need to put the new command withot counter?

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java (hunk 1 above)
Comment (YHerm) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java:14]: I didn't found a better name and I think its a fair name cause all the class is doing is generating the command path and return it, butt if you have a better name idea you can suggest.

Comment (Stormageddon37) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java:14]: My point this class has a name that sounds like a function (verb + noun)

How about PathGenerator? I think it's just a bit better

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java (hunk 2 above)
Comment (Stormageddon37) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java:28]: You're right, foreach won't work

I imagined something like this:

```java
GBCommand[] nodeCommands = IntStream.range(0, path.size() - 1)
                                    .mapToObj(i -> new StraightLineMotion(path.get(i), path.get(i + 1)))
                                    .toArray(GBCommand[]::new);
```

File: src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java (hunk 1 above)
Comment (YHerm) [reply src/main/java/edu/greenblitz/tobyDetermined/commands/telescopicArm/moveByNodes/GeneratePath.java:14]: Ok

File: src/main/java/edu/greenblitz/utils/GBMath.java [hunk 14]
```diff
@@ -1,20 +1,108 @@
...
-	 * @param y the modulo base
-	 * @return the correct modulo result
-	 */
-	public static double absoluteModulo(double x, double y) {
-		return ((x % y) + y) % y;
-	}
-	
-	
+
+    public static double limit(double velocity, double maxMagnitude) {
```
Comment (tnrz135) [inline src/main/java/edu/greenblitz/utils/GBMath.java:8]: Since this function is in a utils class, it should have more generic names for the general purpose of this function.
Also, you don't have to change this part, but this function is usually called 'clamp'

File: src/main/java/edu/greenblitz/utils/GBMath.java [hunk 15]
```diff
@@ -1,20 +1,108 @@
...
+        return limit(velocity, -maxMagnitude, maxMagnitude);
+    }
+
+    public static double limit(double value, double min, double max) {
+        return Math.min(max, Math.max(min, value));
+    }
+
+    public static double sigNum(double end, double start){
```
Comment (tnrz135) [inline src/main/java/edu/greenblitz/utils/GBMath.java:16]: I know that the name in Java is sigNum, but it is not clear enough

File: src/main/java/edu/greenblitz/utils/GBMath.java [hunk 16]
```diff
@@ -1,20 +1,108 @@
...
+
+    /**
+     * A useful interpolation function
+     *
+     * @param x          the variable of the function
+     * @param size       a constant scaling of the equation
+     * @param uniformity a kind of uniformity term
+     * @param xMovement  a translation of the function to the left by xMovement units
+     * @return a Pair which houses the result in this format (length,angle)
+     */
```
Comment (tnrz135) [inline src/main/java/edu/greenblitz/utils/GBMath.java:92]: In other use in the code you gave a link explaining what is sigmoid, I think you should add it here too.

File: src/main/java/edu/greenblitz/utils/GBMath.java [hunk 17]
```diff
@@ -1,20 +1,108 @@
...
+    public static double sigmoid(double x, double size, double uniformity, double xMovement) {
+        return size / (1 + Math.exp(uniformity * (xMovement - x)));
+    }
+
+    public static double lawOfCosines(double sideA, double sideB, double angleBetweenSideAndSideB) {
+        return Math.sqrt(sideA * sideA + sideB * sideB - 2 * sideA * sideB * Math.cos(angleBetweenSideAndSideB));
+    }
+
+    public static double getBow(double length1, double angle1, double angle2) {
```
Comment (tnrz135) [inline src/main/java/edu/greenblitz/utils/GBMath.java:101]: I got the bow, what next?
image

**Review by tnrz135 — DISMISSED** [review]

File: src/main/java/edu/greenblitz/utils/GBMath.java (hunk 14 above)
Comment (YHerm) [reply src/main/java/edu/greenblitz/utils/GBMath.java:8]: but the function return the value that it gets after limitation so what can be more generic name?

**Review by tnrz135 — APPROVED** [review]

**Review by Stormageddon37 — APPROVED** [review]
